Package authoring best practices guide (#2166)
NuGet/docs.microsoft.com-nuget: docs/guides/Package-authoring-best-practices.md

---
title: Package authoring best practices
description: A general guide of best practices for creating high quality NuGet packages.
author: chgill-MSFT
ms.author: chgill
ms.date: 09/17/2020
ms.topic: conceptual
---

# Package authoring best practices

> This guidance is intended to give NuGet.org package authors a lightweight end to end reference for creating and publishing high quality packages. It will primarily focus on package-specific best practices such as metadata and packing. For more in-depth suggestions for building high quality libraries, see the .NET [Open-source library guidance](https://docs.microsoft.com/dotnet/standard/library-guidance/).

Reviewer: What do you think of this?

```suggestion
This guidance is intended to give NuGet.org package authors a lightweight reference to create and publish high-quality packages. It will primarily focus on package-specific best practices such as metadata and packing. For more in-depth suggestions for building high quality libraries, see the .NET [Open-source library guidance](https://docs.microsoft.com/dotnet/standard/library-guidance/).
```

Create a package from your SDK-style project by defining the required properties and packing in [Visual Studio](https://docs.microsoft.com/nuget/quickstart/create-and-publish-a-package-using-visual-studio?tabs=netcore-cli) or the [dotnet CLI](https://docs.microsoft.com/nuget/quickstart/create-and-publish-a-package-using-the-dotnet-cli).

✔️ DO create an SDK-style project and create (pack) your package using Visual Studio or the dotnet CLI.

For more detailed guidance regarding package creation including necessary client tools, project file example, and commands, see [Create a NuGet package using the dotnet CLI](https://docs.microsoft.com/nuget/create-packages/creating-a-package-dotnet-cli).

To help decide which .NET frameworks to target, see our [latest guidance for cross-platform targeting](https://docs.microsoft.com/dotnet/standard/library-guidance/cross-platform-targeting).

## Package metadata

> Metadata is a foundational component of any NuGet package. The quality of your metadata can vastly influence the discoverability, usability, adn trustworthiness of your package.

Reviewer: ```suggestion
Metadata is a foundational component of any NuGet package. The quality of your metadata can vastly influence the discoverability, usability, and trustworthiness of your package.
```

In Visual Studio, the recommended way to specify package metadata is to go Project > [Project Name] Properties > Package.

Package metadata elements can also be [specified directly in the project file](https://docs.microsoft.com/nuget/create-packages/creating-a-package-msbuild#set-properties).

> Below is a table mapping and describing available packages metadata elements:

Reviewer 2: @nkolev92 good point. I think a subset of https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#pack-target is acceptable so long as we maintain it as a best practice rather than a tallied list of properties added to pack overtime.

> Below is a table mapping and describing available packages metadata elements:

Reviewer: ```suggestion
Below is a table mapping and describing available package metadata elements:
```

| [`Release notes`](#release-notes)             | [`PackageReleaseNotes`](https://docs.microsoft.com/dotnet/core/tools/csproj#packagereleasenotes)                                          | [`releaseNotes`](https://docs.microsoft.com/nuget/reference/nuspec#releasenotes)                  | A description of the changes made in this release of the package.                                                 |  |

### Package ID

If you're publishing a completely new package:

✔️ DO choose a package ID that is unique and clearly differentiated from existing packages on NuGet.org.
> You can check if a package ID is unique and differentiable by searching for the ID on NuGet.org or checking if the following link exists: https://www.nuget.org/packages/<package name\>.

✔️ CONSIDER choosing a NuGet package name with a prefix that meets NuGet's [prefix reservation criteria](https://docs.microsoft.com/nuget/nuget-org/id-prefix-reservation#id-prefix-reservation-criteria).
> Reserving the prefix ID for your package will let you get the verified check mark:
> ![image](media/Verified-check-mark.png)

> > Check out the [Package ID prefix reservation docs](https://docs.microsoft.com/nuget/nuget-org/id-prefix-reservation) to learn more.

Reviewer: This is wrapping weirdly on the docs page:

![image](https://user-images.githubusercontent.com/737941/106338792-8ea14380-6249-11eb-9a9b-4c342f8720e2.png)

I'd consider adding a new line after the image

Author: ```suggestion
>
> Check out the [Package ID prefix reservation docs](https://docs.microsoft.com/nuget/nuget-org/id-prefix-reservation) to learn more.
```

### Package Version

✔️ CONSIDER using [SemVer](https://semver.org/) to version your NuGet package.
> Essentially, this means using the Major.Minor.Patch[-prerelease] format.

✔️ DO publish a package as a [pre-release package](https://docs.microsoft.com/dotnet/standard/library-guidance/nuget#pre-release-packages) if it is non-stable or a preview.

> See the [.NET library versioning guide](https://docs.microsoft.com/dotnet/standard/library-guidance/versioning) for more advanced guidance including assembly versioning.

Reviewer: Consider dropping the `assembly versioning` bit. This is only important if you strong name your assemblies, which is a less common these days (see [this](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/strong-naming#create-strong-named-net-libraries)).

```suggestion
See the [.NET library versioning guide](https://docs.microsoft.com/dotnet/standard/library-guidance/versioning) for more advanced guidance.
```

### Authors

✔️ DO use the author field for your or your organization's "pretty name."
> For example, if my NuGet.org username is "jdoe" then using "Jane Doe" for the author field may help consumers recognize me as an author. If my organization's NuGet.org username is "ContosoToolkit" then using "Contoso Corporation" may be more recognizable and inspire more consumer trust.
### Description

✔️ DO include a short description (up to 4000 characters) to describe your package.

> > Package descriptions are one of the most prominent fields surfaced in NuGet search and will likely be the first thing a potential consumers looks at to determine if a package is right for them.

Reviewer: ```suggestion
> Package descriptions are one of the most prominent fields surfaced in NuGet search and will likely be the first thing potential consumers looks at to determine if a package is right for them.
```

---
title: Package authoring best practices
description: A general guide of best practices for creating high quality NuGet packages.
author: chgill-MSFT
ms.author: chgill
ms.date: 09/17/20
ms.topic: conceptual
---

> # Package authoring best practices

Reviewer 2: Overall I really like the structure of this doc! Could we also add more visuals as to what we are referring to? I think this would especially help users know "Why" it's important.

Reviewer 3: Not sure if this is what you meant @JonDouglas, but I wanted to +1 the "Why" part. Could we give a short motivation behind **ALL** the best practices? We have them in many cases but not all.

Author: Definitely plan to add more visual examples. Also agree that the "Why" need to be more upfront and clear.

Reviewer 4: Can we have a sample package with all the best practices there? http://github.com/nuget/samples 

Then it may be useful to refer to the actual line for reference as part of these metadata suggestions as required.

> # Package authoring best practices

Reviewer 5: Should we list some of the most important things that NuGet already has pack warnings about?  For example, things related to PackageReference compatibility. I'm not 100% sure we have a pack warning for all of these but:
* tools\install.ps1 and tols\uninstall.ps1 are not compatible with PackageReference
* If using `content\`, also include `contentFiles\`
* If adding msbuild props\targets to your package, understand `build\` vs `buildTransitive\`, and prefer `build\<tfm>\packageid.props` rather than `build\packageid.props`.

Reviewer 2: I think we should also consider anything that is a best practice & doesn't warn today as a way we can also enhance the packing experience.

> This guidance is intended to give NuGet.org package authors a light-weight end to end reference for creating and publishing high quality packages. It will primarily focus on package-specific best practices such as metadata and packing. For more in depth suggestions for building high quality libraries, see the .NET [Open-source library guidance](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/).

Reviewer 3: Maybe a "CONSIDER" on author package signing?

> This guidance is intended to give NuGet.org package authors a light-weight end to end reference for creating and publishing high quality packages. It will primarily focus on package-specific best practices such as metadata and packing. For more in depth suggestions for building high quality libraries, see the .NET [Open-source library guidance](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/).

Reviewer 3: Is there any recommendation around minClientVersion? Or do we only want people that "know what they're doing" to use it?

> This guidance is intended to give NuGet.org package authors a light-weight end to end reference for creating and publishing high quality packages. It will primarily focus on package-specific best practices such as metadata and packing. For more in depth suggestions for building high quality libraries, see the .NET [Open-source library guidance](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/).

Reviewer 3: I feel like there's a notion floating in the air that CI builds should not go to nuget.org but only important, significant releases (prerelease or stable, both are fine). Do you want a recommendation around that? CI builds make should go to GPR/AzDO/MyGet and official releases go to nuget.org.

> This guidance is intended to give NuGet.org package authors a light-weight end to end reference for creating and publishing high quality packages. It will primarily focus on package-specific best practices such as metadata and packing. For more in depth suggestions for building high quality libraries, see the .NET [Open-source library guidance](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/).

Reviewer 5: should we rename the file, or move it to the server team sub-url, if it's nuget.org specific?  Should we split the guidelines into generic package authoring guidelines, and nuget.org specific ones?

Author: Would there be significant differences between guidelines for NuGet.org vs private packages?

Reviewer 5: from a quick look:
* checking if a package id is used on nuget.org is unlikely to be relevant
* nuget.org's prefix reservation criteria is definitely not relevant
* authors matching the nuget.org account doesn't make sense
* license might not make sense/be relevant
* icon is still good advice, but probably much less likely and less important on a private feed than nuget.org. Particularly when an enterprise has a private feed per division/product/team.
* release notes may still be good advice, but it's generally really hard to do release notes, so in my opinion unlikely to actually get done.  In my previous team we pushed a new package after every commit, so the release notes would just be a duplicate of the git log. And often these packages weren't even for other teams to use, but a way to share code between different repos that make up a larger system (much like how nuget.org's source is split across multiple repos)

That's at least half the recommendations in this PR :)  I would hope that enterprise customers creating packages for internal feeds would think about each one and decide what's relevant to them. And for general best practices, maybe we should have pack warnings rather than docs, although that's very tricky because any new warning breaks people who treat warnings as errors, and they don't care that they asked for it, they still get angry about anything that breaks their build.

I'm not pushing for this, but it's a thought of mine. Several of the recommendations are nuget.org specific, and since nuget.org docs have their own subsection, does it below there?  If an enterprise customer is creating a package for proprietary, internal-only use, will they read the first few recommendations, see it's nuget.org specific, and stop reading the rest?   Or am I overthinking this and it's all fine as is?

> This guidance is intended to give NuGet.org package authors a light-weight end to end reference for creating and publishing high quality packages. It will primarily focus on package-specific best practices such as metadata and packing. For more in depth suggestions for building high quality libraries, see the .NET [Open-source library guidance](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/).

Reviewer 3: Are there recommendations around the "requireLicenseAcceptance" and "language" property?

> This guidance is intended to give NuGet.org package authors a light-weight end to end reference for creating and publishing high quality packages. It will primarily focus on package-specific best practices such as metadata and packing. For more in depth suggestions for building high quality libraries, see the .NET [Open-source library guidance](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/).

Reviewer 3: Is `<frameworkAssemblies>` even interesting anymore? Should we have an AVOID? I've always felt that I only want a project to bring in project references and I'll select my direct assembly references myself.

Reviewer 5: I think it's still relevant. Even though most of the BCL is available as packages now, by default a project targeting .NET Framework will not reference `System.Net.Http`, so a customer wanting to use `HttpClient` will either need to reference the assembly, or the package. Referencing the package requires a restore, which is not needed for an assembly reference. Referencing the package introduces all sorts of risks of package downgrade warnings, and I've seen some assembly version issues as well (MSBuild warnings in Resolve Assembly References), whereas assembly references seem to have fewer issues.

Or maybe this is just my personal preference.  But to me, as long as .NET Framework is in use, I will always prefer assembly references over packages.

Reviewer 3: Okay, sounds good! Maybe this would be a good "CONSIDER" then 😄

> This guidance is intended to give NuGet.org package authors a light-weight end to end reference for creating and publishing high quality packages. It will primarily focus on package-specific best practices such as metadata and packing. For more in depth suggestions for building high quality libraries, see the .NET [Open-source library guidance](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/).

Reviewer: ```suggestion
This guidance is intended to give NuGet.org package authors a lightweight end to end reference for creating and publishing high quality packages. It will primarily focus on package-specific best practices such as metadata and packing. For more in depth suggestions for building high quality libraries, see the .NET [Open-source library guidance](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/).
```

## Types of recommendations

> Each article presents four types of recommendations: **Do**, **Consider**, **Avoid**, and **Do not**. The type of recommendation indicates how strongly it should be followed.

Reviewer 3: ```suggestion
Each article presents four types of recommendations: **Do**, **Consider**, **Avoid**, and **Do not**. The type of recommendation indicates how closely it should be followed.
```

You should almost always follow a **Do** recommendation. For example:

✔️ DO include a short description for your package that describes what it's for.

> On the other hand, **Consider** recommendations should generally be followed, but there are legitimate exceptions to the rule and you shouldn't feel bad about not following the guidance:

Reviewer: The last segment is a little redundant, what do you think of this?

```suggestion
On the other hand, **Consider** recommendations should generally be followed, but there are legitimate exceptions to the rule:
```

✔️ CONSIDER choosing a NuGet package name with a prefix that meets NuGet's prefix reservation [criteria](https://docs.microsoft.com/en-us/nuget/reference/id-prefix-reservation).

**Avoid** recommendations mention things that are generally not a good idea, but breaking the rule sometimes makes sense:

❌ AVOID NuGet package references that demand an exact version.

And finally, **Do not** recommendations indicate something you should almost never do:

> ❌ DO NOT use the `LicenseUrl` metadataproperty.

Reviewer 3: ```suggestion
❌ DO NOT use the `LicenseUrl` metadata property.
```

> 

Reviewer 2: Before frameworks, should we talk about how to pack? We want to push people to SDK-style packing instead of .nuspecs right?

https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/nuget#create-a-nuget-package

Author: My original outline planned to have that at the end since chronologically you pack after you've already written the package. What do you think?

Reviewer 2: I think **where** you put the information is the first thing I'd be doing. Whether that's as XML in a .csproj or a .nuspec.

Reviewer 5: I don't understand what you mean by "you pack after you've already written the package". While in the development cycle pack happens at the end, after you write code, compile and debug, the customer needs a project file in order to do these things. If they start out with a non-SDK style project, then write code, when they come to pack and read these recommendations, if the last thing is says is "hey, it would have been easier if you used an SDK style project", that's not great. I think guiding people to SDK style projects early is better (and this applies to ALL our docs, which are currently far too heavy on non-SDK style projects and custom nuspec files, since they were written before SDK style projects and never updated. But now I'm really getting off-topic.

Also, I don't currently see any guidance on preferring SDK style projects in the current file.

I think it's useful as one of the first recommendations, because even though they're not intended to be in order of importance, culturally often things at top of lists are more "important". Although there's also a case to be made about the last thing read is more likely to stick. However, some people will start reading from the top and not make it all the way to the end, so there might be greater reach by putting it at the top.

Reviewer 6: Remember pack is really split on PackageReference vs non-PackageReference. 

legacy vs sdk has a difference in how you enable the functionality, but that's something that can be fixed.

Highly inclusive packages are cross-platform and compatible with as many frameworks as possible. Today, that means using .NET Standard. .NET Standard is a specification of .NET APIs that are available on all .NET implementations.

✔️ DO start with including a `netstandard2.0` target.
> Most general-purpose libraries should not need APIs outside of .NET Standard 2.0. .NET Standard 2.0 is supported by all modern platforms and is the recommended way to support multiple platforms with one target.

See the .NET [Cross-platform targeting guidance](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/cross-platform-targeting) for more advanced suggestions regarding cross-platform and multitargeting scenarios.

## Package metadata

Metadata is a foundational component of any NuGet package. The quality of your metadata can vastly influence the discoverability and usability of your package.

In Visual Studio, the recommended way to specify package metadata is to go Project > [Project Name] Properties > Package.

> Package metadata elements can also be specified directly in the project file (.csproj).

Reviewer: Should we be language agnostic and include F#'s .fsproj project files?

Author: TIL F# has an .fsproj file instead of a .csproj file 😂

Reviewer 7: Consider linking to documentation where possible. For example here: https://docs.microsoft.com/en-us/nuget/create-packages/creating-a-package-msbuild#set-properties

> Below is a table mapping and describing all the available packages metadata elements:

Reviewer: I would avoid committing to an exhaustive guide here :)

```suggestion
Below is a table mapping and describing the available packages metadata elements:
```

> | Visual Studio property name | csproj/MSBuild property name | Nuspec property name        | Description                                                                                                       |

Reviewer 3: Why is this mapping table here? It doesn't have much to do with best practices themselves, right? Wouldn't this be better off in its own document?

Reviewer 5: [It already is](https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#pack-target), although in a place that may not be obvious to many people. But that's a separate issue. Since you pointed it out, I agree, it doesn't seem like it should be in the recommended practises document.

Author: I'm not sure I agree. 

We don't go into detail on how to specify all these properties in the [Create a package documentation](https://docs.microsoft.com/en-us/nuget/create-packages/creating-a-package-dotnet-cli). Someone reading this shouldn't have to look up a reference doc to to implement the suggestions. 

The table @zivkan linked also doesn't include Visual Studio property fields in the mapping (which may be more useful to beginners). 

Toward the goal of making this document as actionable as possible, do you think there is a more useful way we could present this information?

Reviewer 5: Documentation on how to modify package metadata is useful not only to a best practices document, but to all customers creating packages. Therefore, if the "visual studio property fields" is useful, it should be available somewhere under the "authoring packages" section.

About making this document as actionable as possible, my feedback is that my personal preference is that the document contains information relevant to the topic at hand, which is recommendations. If I'm intermediate to experienced with NuGet, reading a document that duplicates beginner information will make me lose interest quickly, and I might miss the primary information the documentation author intended for me to learn.  But I understand that this is not as beginner friendly. You're the PM, so I'll leave it up to you where you want the docs to end up.

Something else I'll mention is that the more information we duplicate, the higher the risk that the docs get outdated/wrong if something in the product changes and only some of the duplicated docs are changed.

Reviewer: @zivkan brings up good points. Personally, I'd vote to link to the relevant metadata docs for each recommendation.

Author: @loic-sharma @JonDouglas thoughts on including the whole table vs. linking to it in another doc?

Reviewer 4: My vote is to put the fields/metadata referred in this doc here and a link to the complete (?) set of metadata towards the end. The best practices guide would not mention all the fields available for authors  and as we make updates to it, we can remember to update it here.

Reviewer 6: Documentation duplication scares me so much. 
It's very easy to end up in a situation where one thing is updated, while the other one isn't. 

Is there a way to integrate parts of other docs into this one? 
I know there was a way to do it with the code snippets that we did for the nuget-sdk samples.

| `Licensing - Expression`    | `PackageLicenseExpression`   | `license type="expression"` | An SPDX license expression or path to a license file within the package, often shown in UIs like NuGet.org.       |
| `Licensing - File`          | `PackageLicenseFile`         | `license type="file"`       | File for custom license that isn't supported by license expressions                                               |
| `Project URL`               | `PackageProjectUrl`          | `projectUrl`                | A URL for the project homepage.                                                                                   |
| `Icon File`                 | `PackageIcon`                | `icon`                      | An image to use as the icon for the package.                                                                      |
| `Repository URL`            | `RepositoryUrl`              | `repository url=""`         | URL to the repository from which the package was built.                                                           |

> | `Repository type`           | `RespositoryType`            | `repository type=""`        | Type of repository the repository URL is pointing to (i.e. git)                                                   |

Reviewer 5: I'd suggest that customers packing via msbuild (dotnet cli uses msbuild) would be better off using one of the SourceLink packages, which automatically adds the repository type and url metadata.

edit: ah, I see you're added it further down. Maybe it's ok as it is now, but also consider if customers are going to read this, stop reading, go to VS and make edits, and then come back to read more of the guidelines.

| `Tags`                      | `PackageTags`                | `tags`                      | A space-delimited list of tags and keywords that describe the package. Tags are used when searching for packages. |
| `Release notes`             | `PackageReleaseNotes`        | `releaseNotes`              | A description of the changes made in this release of the package.                                                 |

### Package ID

If you're publishing a completely new package:

✔️ DO choose a package ID that is unique and clearly differentiated from existing packages on NuGet.org.

> > You can make sure a package ID is unique and differntiable by searching for the ID on NuGet.org or checking if the following link exists: https://www.nuget.org/packages/<package name\>.

Reviewer 5: ```suggestion
> You can check if a package ID is unique and differentiable by searching for the ID on NuGet.org or checking if the following link exists: https://www.nuget.org/packages/<package name\>.
```

I don't like the working "make sure", because as several people mentioned in the Teams chat, there are uncommon scenarios where this won't work. I think "check" makes it less committal that this is guaranteed.

Reviewer 2: differentiable -> unique?

Reviewer 2: Sorry I'm reading "unique and unique" Why do we have to say differentiable? Unique is the same word.

> ✔️ CONSIDER choosing a NuGet package name with a prefix that meets NuGet's prefix reservation [criteria](https://docs.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation#id-prefix-reservation-criteria).

Reviewer: Consider making more of the text a link here since it matches the section heading:

```suggestion
✔️ CONSIDER choosing a NuGet package name with a prefix that meets NuGet's [prefix reservation criteria](https://docs.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation#id-prefix-reservation-criteria).
```

> Reserving the prefix ID for your package will let you get the verified check mark:
> ![image](media/Verified-check-mark.png)
> Check out the [Package ID prefix reservation docs](https://docs.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation) to learn more.

### Package Version

> ✔️ CONSIDER using [SemVer 2.0.0](https://semver.org/) to version your NuGet package.

Reviewer 3: I would recommend just saying "Semantic Versions (SemVer)" without specifying 2.0.0 at first. There is a lot of other references in docs to SemVer 2.0.0 in light of how it is different than SemVer 1.0.0 (dots in pre-release and build metadata). Consider mentioning something like "NuGet supports the SemVer 2.0.0 specification, however some older clients may only support SemVer 1.0.0" with a link to the distinctions.

Reviewer: Just curious, what's the value in this recommendation? Off-hand this appears superfluous, should we remove it?

Author: My understanding was that SemVer is a common industry standard, although not universally adopted on NuGet.org. Doesn't encouraging SemVer help created shared understanding around versioning meaning (i.e. Major.Minor.Patch)? I'm far from an expert though, so I'm happy to hear any specific critiques 😊

Reviewer 4: I am a bit bit confused with "CONSIDER". If these are best practices guide, shouldn't we just say "Use.."? Why leave the iota of doubt? 😄. The rationale and explanation should be enough for authors to decide whether its something they should use or not.

Reviewer 6: I think this should be `Do` as well. 

Have given collapsible sections for explanations a thought? Markdown is pretty limiting as far as that goes and I have no  clue if it's a docs recommendation.

Reviewer: What's our goal in recommending semver? Do we want people to version their packages properly to declare breaking changes, new features, and bug fixes? If so, we should clarify the "why" of semver. If not, what's the value in recommending semver? That folks avoid using 4 digits to represent their versions? That doesn't seem all that useful.

> Essentially, this means using the Major.Minor.Patch[-prerelease] format.

> ✔️ DO include a pre-release suffix when releasing a non-stable package.

Reviewer 2: https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/nuget#pre-release-packages also has some information we could use?

> See the [.NET library versioning guide](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/versioning) for more advanced guidance including aseembly versioning.

Reviewer 3: Should we have a recommendation against the 4th version digit since it is not strictly SemVer-compliant?

Author: You mean the 4th digit mentioned in the .NET docs?

Reviewer 3: I'm not sure of the specific spot that you're referring to. What I mean is that NuGet supports a version 1.0.0.1 but this is not technically SemVer compliant. SemVer only supports 1.0.0 (three digits).

> See the [.NET library versioning guide](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/versioning) for more advanced guidance including aseembly versioning.

Author: ```suggestion
See the [.NET library versioning guide](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/versioning) for more advanced guidance including assembly versioning.
```

### Authors

> ✔️ CONSIDER including the current NuGet.org owners for consistency across NuGet clients.

Reviewer 3: I don't think I agree with this one. Authors is more generally used for a pretty human or company name where as username is often abbreviated and lowercase.
![image](https://user-images.githubusercontent.com/94054/93631020-c77a0380-f99f-11ea-943c-a3bdeede19f8.png)

Author: That's true, but currently we require consistency between Owner, Author, and prefix for prefix reservations. So it'd be weird that that's not a "best practice." It's a bit of a band-aid for not showing Owners in the PMUI Browse tab, and I agree that it's kind of weird.
@anangaur @JonDouglas thoughts on this one? Personally, I was a big conflicted.

Reviewer 4: Author is used for 2 purposes today:
1. To use a better nam, may be trademarked one. E.g. `My Foo Company Limited` corresponding to owner = `myfoo`
2. To credit the actual author(s) that's different from owner. e.g. `Microsoft and Collaborators` or `<Person name>`

IMO we could say make it the same as the owner with a pretty name. Thoughts?

Author: You mean like

Authors: NuGet, Chris Gill

?

Author: So have both Owners + Pretty Name?

Reviewer 3: Using it as a band-aid for showing Owners in the PMUI is not great. First, people can lie (putting `Microsoft` in the authors field), so any emphasis on authors == owners or owners == authors is a bad signal to both package authors and consumers. Second, owners can change over time. For example, Newtonsoft.Json used to only be owned by `jamesnk` but not it's owned by a new organization and .NET Foundation.

The author should be use _authored_ the package. It's like a book publisher vs. book author (in my mind). nuget.org is the book store in that analogy.

Reviewer 3: @anangaur, I think a trademarked and/or pretty name for the author is good. So for Microsoft packages it's Microsoft but for Newtonsoft.Json it's James Newton-King.

Reviewer 6: I agree with @joelverhagen. 

This recommendation falls apart quickly when the author is an individual and not a company.

> Visual Studio currently displayed `Authors` on the Browse tab. It is also in the criteria for ID prefix reservation.

### Description

> ✔️ DO include a short description for your package that describes what it's for.

Reviewer 3: Are there recommendations on max length? I think nuget.org caps it at 4000 but that's perhaps too long for many experiences like search results in VS.

Reviewer 5: Does nuget.org have docs that list all their package limits? (package id limit, version string limit, etc)?  Maybe a recommendation should be to check the limits for the source you plan to puhs your package to, and here's the docs for nuget.org?  although, many of the recommendations are not so important for enterprise customers pushing to an internal-only feed.

Reviewer 3: The limits are listed on the .nuspec reference page, e.g. https://docs.microsoft.com/en-us/nuget/reference/nuspec#description

Author: Length suggestion is a good add I think 👍

Reviewer 4: **Short** LGTM. May be revisit this once we have support for README?

> Package descriptions are one of the most prominent fields surfaced in NuGet search and will likely be the first thing a potential consumers looks at to determine if a package is right for them.

### Copyright

✔️ CONSIDER copyrighting your package with "Copyright (c) <name/company\> <year\>."

> >A copyright notice essentially indicates that your work cannot be copied without your permission. Including, a copyright notice in your package is easy and won't do any harm!

Reviewer: ```suggestion
>A copyright notice essentially indicates that your work cannot be copied without your permission. Including a copyright notice in your package is easy and won't do any harm!
```

Just double-checking - is this recommendation lawyer approved? :)

Reviewer: Just double-checking - is this recommendation lawyer approved? :)

Author: It is not, is Adam on GitHub? 😁

Example: Copyright (c) Contoso 2020

### Licensing

✔️ DO specify a valid license for your package.
> [!IMPORTANT]
> A project without a license defaults to [exclusive copyright](https://choosealicense.com/no-permission/), making it legally impossible for other people to use.

> ✔️ CONSIDER using a license expression (SPDX identifier).

Reviewer 3: Would it be bad to use "Avoid" for license files? In general the best experience would be when users SPDX instead of taking the equivalent SPDX license text and putting it in a file. I think we should encourage people down that path as strongly as possible. SPDX allows for futures of filtering, grouping, etc. A license file is not very machine readable (although there is software that can do it with varying success).

Author: Personally, I'm wary of telling people to Avoid the proper use of a feature we continue to clearly support. I also think it may be redundant with with the suggestion to use license expression when possible.

In general, it's already more work for the user to include a license file over using a license expression. But in scenarios where a company is very legally cautious (license expression doesn't support signed licenses that are specific to the company/package) or where they have a custom license, I don't want to imply that having a license file is a "bad" practice.

Thoughts?

Reviewer 3: Sounds good to me.

> License expressions are surfaced the most clearly to package consumers and make it more obvious if the license has changed with a new release.

> ✔️ CONSIDER specifying an MIT license.

Reviewer 2: A link to https://choosealicense.com/ would be nice somewhere. This is a great place for it!

Reviewer 4: @JonDouglas is this a known OSS place for redirect? Do we know if this URL can be trusted?

Reviewer 6: This one doesn't sit right for me :) 
This implies the package is OSS. Suggesting actual license usage is a big no-no I think.

cc @zkat

Reviewer 2: I'm no expert in OSS licenses. MIT is by far most popular on GitHub with the whole set it and forget it approach. However GPL (copyleft) licenses can be more beneficial to OSS in the long run. Regardless, this website answers all that for somebody. Let's not reinvent the wheel.

To me it might say:

✔️ CONSIDER [choosing an open source license](https://choosealicense.com/) if your project is open source.

Reviewer 2: @anangaur https://github.com/github/choosealicense.com

> If you want your package as to be usable by as many consumers as possible, the MIT license is a very well known non-restrictive license.

> ❌ DO NOT use the deprecated `LicenseUrl` metadataproperty.

Reviewer: Do all servers like AzDO support the new license metadata fields? If not, should we restrict this recommendation to nuget.org? Or, should we change this to `AVOID`?

> ❌ DO NOT use the deprecated `LicenseUrl` metadataproperty.

Reviewer: ```suggestion
❌ DO NOT use the deprecated `LicenseUrl` metadata property.
```

> This presents legal ambiguity as license changes at the URL will retroactively change the displayed license for previous package versions.

### Project URL

✔️ CONSIDER including a link to an associated project or company website.
> Company website links, even if not specific to the package/project, still improve consumer trust.

> ❌ DO NOT put a repository URL (GitHub, BitBucket, etc.) in the `Project URL` field, as consumers will expect to find it in the `Repository URL` field.

Reviewer 3: I would make this an "Avoid" not a "DO NOT". Do we have confidence that repository URL has good support on all tooling (both 1st and 3rd party)?

Author: Could you elaborate a bit further? Any specific tools you're worried about?

Reviewer 3: Well, VS PM UI does not show repository URL, IIRC. And for many small open source projects the project URL _is_ the repository URL.

Author: TIL PMUI doesn't show repo URL. Why don't we?

Reviewer 3: The work was never done. The metadata is not plumbed through to all of the API endpoints that need it (package metadata resource, i.e. registration) and the client work was not done.
https://github.com/NuGet/Home/issues/8584

### Icon

✔️ CONSIDER including an icon with your package to help visually differentiate it. It's a relatively small addition that can improve perception of package quality.
> Icons can be specific to individual packages or be a brand logo.

✔️ DO use an image that is 128x128 and has a transparent background (PNG) for best viewing results.
> NuGet will automatically scale your image to the client it is being displayed on.

❌ DO NOT use the deprecated `IconUrl` metadata property.

### Repository Type and URL

> ✔️ CONSIDER including a repository type and URL to make it easy for package consumers to view sourecode and package documentation (only if public).

Reviewer 3: ```suggestion
✔️ CONSIDER including a repository type and URL to make it easy for package consumers to view soure code and package documentation (only if public).
```

> ✔️ CONSIDER setting up [Source Link](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/sourcelink) to add source control metadata to your assemblies and NuGet package.

Reviewer 2: I think we should expand on this:

https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/nuget#symbol-packages

And this doc?

https://docs.microsoft.com/en-us/nuget/create-packages/symbol-packages-snupkg

Reviewer 3: Yes, I think calling out "snupkg"/symbols packages is important. Source link is one path but it's not the only one and if a user can't user Source Link for some reason, consider still publishing a snupkg with, say, embedded sources.

And what about the snupkg vs. PDB in the .nupkg discussion. Any recommendations around that?

Reviewer: Our official recommendation has been to put PDBs in the .snupkg (to reduce .nupkg size), but this has been controversial as many argue that embedded PDBs provide a simpler experience.

Author: What's the experience for including PDBs in the package vs in a .snupkg?

I always just add this:
```
<PropertyGroup>
    <IncludeSymbols>true</IncludeSymbols>
    <SymbolPackageFormat>snupkg</SymbolPackageFormat>
</PropertyGroup>
```

Reviewer 6: If we're talking snupkgs here we need to call out the portable pdb caveat.

Reviewer 2: I've heard the argument of snupkg vs embedded symbols so many times that I don't think there is an official recommendation. Let's just put in both scenarios for now until that's clearer by the .NET team / etc?

> Source Link will automatically adds `Repository URL` and `Repository Type` to the package metadata. It also adds the specific commit associated with your package version.

### Tags

> ✔️ DO include several tags with key terms related to your package to enhance discoverability.

Reviewer 2: Here's an example:

`logging`, `mocking`, `unittest`, `database`, `storage`

> Tags are taken into account in NuGet's search algorithm and are especially helpful for terms that are not in the Package ID.

### Release notes

✔️ CONSIDER including release notes with each update describing what changes were made.
> While there is no specific format required for release notes, we recommend including:
>
> 1. Breaking changes
> 2. New features
> 3. Bug fixes
> 
> If you already track release notes in your repo, you can also include a link to your release notes file.

> ## Dependencies

Reviewer 5: Customers using "generated" pack can't, but customers specifying their own nuspec file can create dependencies directly under the nuspec's dependency node, rather than creating a group per supported TFM. nuget will warn on pack when it detects tfm-specific assets and no dependency group for that tfm, but should we also have a note in here saying to do that?  But as mentioned, anyone who lets nuget generate the nuspec for them will never have this problem.

Reviewer 6: This is great. 

I'd love to see a `DO NOT use the dependencies element`. 

Throw in a couple of `do not` for good measure.

Reviewer: Would it be better to recommend packing using proj files (instead of .nuspec files)? To me, generating a package using a .nuspec file is now an "advanced" scenario.

### Lean packages

✔️ DO review your package for unnecessary dependencies.
> Unnecessary dependencies needlessly increase the size of your package as well as increase likelihood of version conflicts for both yourself and your consumers!

> ### Depencency version ranges

Reviewer: ```suggestion
### Dependency version ranges
```

## Types of recommendations

Each article presents four types of recommendations: **Do**, **Consider**, **Avoid**, and **Do not**. The type of recommendation indicates how closely it should be followed.

You should almost always follow a **Do** recommendation. For example:

✔️ DO include a short description for your package that describes what it's for.

On the other hand, **Consider** recommendations should generally be followed, but there are legitimate exceptions to the rule:

✔️ CONSIDER choosing a NuGet package name with a prefix that meets NuGet's prefix reservation [criteria](https://docs.microsoft.com/nuget/reference/id-prefix-reservation).

**Avoid** recommendations mention things that are generally not a good idea, but breaking the rule sometimes makes sense:

❌ AVOID NuGet package references that demand an exact version.

And finally, **Do not** recommendations indicate something you should almost never do:

❌ DO NOT use the `LicenseUrl` metadata property.

## Create a NuGet package

> The latest recommended way to to create a NuGet package is from an [SDK-style project](https://docs.microsoft.com/nuget/resources/check-project-format). SDK-style project properties, including target [framework](#framework-targeting) and [package metadata](#package-metadata), are defined in the project file (.csproj).

Reviewer: `.csproj` is C# specific. Can we say `(.csproj or .fsproj)` instead? (I've never created an F# package before, please double-check this is correct)

Reviewer 2: Can we simplify and just say ".NET SDK-style projects"?

Reviewer 2: or link to https://docs.microsoft.com/en-us/visualstudio/ide/solutions-and-projects-in-visual-studio?view=vs-2019#project-file

Reviewer 2: there's too many project files to name here. .csproj is the most used, but i think sdk-style project files is most correct.

Author: ```suggestion
The latest recommended way to to create a NuGet package is from an [SDK-style project](https://docs.microsoft.com/nuget/resources/check-project-format). SDK-style project properties, including target [framework](#framework-targeting) and [package metadata](#package-metadata), are defined in the [project file](https://docs.microsoft.com/visualstudio/ide/solutions-and-projects-in-visual-studio?view=vs-2019#project-file).
```

Create package from your SDK-style project by defining the required properties and packing in [Visual Studio](https://docs.microsoft.com/nuget/quickstart/create-and-publish-a-package-using-visual-studio?tabs=netcore-cli) or the [dotnet CLI](https://docs.microsoft.com/nuget/quickstart/create-and-publish-a-package-using-the-dotnet-cli).

✔️ DO create an SDK-style project and create (pack) your package using Visual Studio or the dotnet CLI.

For more detailed guidance regarding package creation including necessary client tools, project file example, and commands, see [Create a NuGet package using the dotnet CLI](https://docs.microsoft.com/nuget/create-packages/creating-a-package-dotnet-cli).
## Framework targeting

Understanding the differences between target frameworks and deciding which one or ones that are right for your library can be very confusing given the wide variety of options available from .NET Framework to .NET 5. So let's try to make it as straightforward as possible.

✔️ CONSIDER starting with [the latest .NET target framework](https://docs.microsoft.com/dotnet/standard/frameworks#latest-versions).

> > Using the latest .NET target framework will allow you to take advantage of the latest .NET technology and language features.

Reviewer: This section is a bit wordy. Also, I'd avoid mentioning .NET 5 directly as otherwise this will need to be updated in a year :)

Also, I disagree with this recommendation. If you target the latest .NET target framework, .NET Framework folks won't be able to use your new package. Personally I would suggest .NET Standard 2.0 if you want wide support, or latest .NET target framework if you want to use latest features. I don't think there's a cookie cutter answer here, this is a decision you'll need to make as author.

Reviewer 2: Definitely agree w/ not mentioning .NET 5 directly. But we should say "the latest .NET version or future version is the foundation for sharing code". 

https://devblogs.microsoft.com/dotnet/the-future-of-net-standard/

I would just point a link to https://docs.microsoft.com/en-us/dotnet/standard/frameworks#latest-versions as they are in the right order of precedence already.

Author: @JonDouglas this section was actually removed entirely based on feedback from Immo. This is all I say now:

![image](https://user-images.githubusercontent.com/15097183/106301406-ae267500-6225-11eb-81fe-db7cd9cc4753.png)

✔️ CONSIDER choosing a NuGet package name with a prefix that meets NuGet's [prefix reservation criteria](https://docs.microsoft.com/nuget/nuget-org/id-prefix-reservation#id-prefix-reservation-criteria).
> Reserving the prefix ID for your package will let you get the verified check mark:
> ![image](media/Verified-check-mark.png)
> Check out the [Package ID prefix reservation docs](https://docs.microsoft.com/nuget/nuget-org/id-prefix-reservation) to learn more.

### Package Version

✔️ CONSIDER using [SemVer](https://semver.org/) to version your NuGet package.
> Essentially, this means using the Major.Minor.Patch[-prerelease] format.

✔️ DO publish a package as a [pre-release package](https://docs.microsoft.com/dotnet/standard/library-guidance/nuget#pre-release-packages) if it is non-stable or a preview.

See the [.NET library versioning guide](https://docs.microsoft.com/dotnet/standard/library-guidance/versioning) for more advanced guidance including assembly versioning.

### Authors

✔️ CONSIDER including the current NuGet.org owners for consistency across NuGet clients.

> > Visual Studio currently displayed `Authors` on the Browse tab. It is also in the criteria for ID prefix reservation.

Reviewer: Should we add a link to the prefix reservation criteria here?

### Description

> ✔️ DO include a short description (up to 4000 characters) for your package that describes what it's for.

Reviewer: What do you think of this?

```suggestion
✔️ DO include a short description (up to 4000 characters) to describe your package.
```

✔️ CONSIDER [including a license expression in your package](https://docs.microsoft.com/nuget/reference/msbuild-targets#packing-a-license-expression-or-a-license-file).
> License expressions are surfaced the most clearly and make it more obvious to consumers if they can use your package or if the license has changed. 
> [!Note]
> NuGet.org only accepts license expressions for licenses that are approved by the Open Source Initiative or the Free Software Foundation.

#### If your package is not open source

✔️ DO [include a license file in your package](https://docs.microsoft.com/nuget/reference/msbuild-targets#packing-a-license-expression-or-a-license-file).
> Any license file (.txt or .md) can be added to your package, including non-standard licenses. 

### Project URL

✔️ CONSIDER including a link to an associated project, repository, or company website.
> Your project site should have everything users need to know about your package and will likely be where users look for documentation.

### Icon

> ✔️ CONSIDER including an icon with your package to help visually differentiate it. It's a relatively small addition that can improve perception of package quality.

Reviewer: Do we have a doc we can link to here on how to embed icons?

> Icons can be specific to individual packages or be a brand logo.

✔️ DO use an image that is 128x128 and has a transparent background (PNG) for best viewing results.
> NuGet will automatically scale your image to the client it is being displayed on.

❌ DO NOT use the deprecated `IconUrl` metadata property.

### Repository Type and URL

> ✔️ CONSIDER including a repository type and URL to make it easy for package consumers to view source code, report issues, and view repository documentation.

Reviewer: Should we put the Source Link paragraph first as it is the best way to fill out this info? I don't feel strongly about this, feel free to keep it as is if you prefer it :)

See the .NET [Cross-platform targeting guidance](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/cross-platform-targeting) for more advanced suggestions regarding cross-platform and multitargeting scenarios.

## Package metadata

Metadata is a foundational component of any NuGet package. The quality of your metadata can vastly influence the discoverability and usability of your package.

In Visual Studio, the recommended way to specify package metadata is to go Project > [Project Name] Properties > Package.

Package metadata elements can also be specified directly in the project file (.csproj).

Below is a table mapping and describing available packages metadata elements:

| Visual Studio property name | csproj/MSBuild property name | Nuspec property name        | Description                                                                                                       |
|-----------------------------|------------------------------|-----------------------------|-------------------------------------------------------------------------------------------------------------------|

> | `Package id`                | `PackageId`                  | `id`                        | The package identifier. A prefix from the identifier can be reserved if it meets the  criteria.                   |

Reviewer 7: On the word "Criteria", please add a link to the criteria docs.  https://docs.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation

Reviewer: FYI you have an extra space:

```suggestion
| `Package id`                | `PackageId`                  | `id`                        | The package identifier. A prefix from the identifier can be reserved if it meets the criteria.                   |
```

> | `Package Version`           | `Version`                    | `version`                   | NuGet package version. For more information, see NuGet package version.                                           |

Reviewer 6: A feedback for later as the doc matures, but items that calls for an action, should ideally link to the area. 

Ex. `see NuGet package version` is weird when the see part is not hyperlinked :)

| `Authors`                   | `Authors`                    | `authors`                   | A comma-separated list of package authors, matching the profile names on NuGet.org.                               |
| `Description`               | `Description`                | `description`               | A long description of the package displayed in UI.                                                                |
| `Copyright`                 | `Copyright`                  | `copyright`                 | Copyright details for the package. language.                                                                      |

> | `Licensing - Expression`    | `PackageLicenseExpression`   | `license type="expression"` | An SPDX license expression or path to a license file within the package, often shown in UIs like NuGet.org.       |

Reviewer 7: Link to documentation: https://docs.microsoft.com/en-us/nuget/reference/nuspec#license

My concern is that a user will not be able to follow those instructions without a lot of bing seaches

### Authors

✔️ CONSIDER including the current NuGet.org owners for consistency across NuGet clients.
> Visual Studio currently displayed `Authors` on the Browse tab. It is also in the criteria for ID prefix reservation.

### Description

✔️ DO include a short description for your package that describes what it's for.
> Package descriptions are one of the most prominent fields surfaced in NuGet search and will likely be the first thing a potential consumers looks at to determine if a package is right for them.

### Copyright

✔️ CONSIDER copyrighting your package with "Copyright (c) <name/company\> <year\>."
>A copyright notice essentially indicates that your work cannot be copied without your permission. Including a copyright notice in your package is easy and won't do any harm!

Example: Copyright (c) Contoso 2020

### Licensing

✔️ DO specify a valid license for your package.
> [!IMPORTANT]
> A project without a license defaults to [exclusive copyright](https://choosealicense.com/no-permission/), making it legally impossible for other people to use.

> ✔️ CONSIDER using a license expression (SPDX identifier).

Reviewer: Should we link to SPDX licenses? https://spdx.org/licenses/

Also, SPDX seems to be a broad term that covers more than just licenses. Consider clarifying:

```suggestion
✔️ CONSIDER using a license expression (SPDX license identifier).
```

Author: ```suggestion
✔️ CONSIDER using a license expression ([SPDX license](https://spdx.org/licenses/) identifier).
```

> [!IMPORTANT]
> A project without a license defaults to [exclusive copyright](https://choosealicense.com/no-permission/), making it legally impossible for other people to use.

❌ DO NOT use the deprecated `LicenseUrl` metadata property.
> This presents legal ambiguity as license changes at the URL will retroactively change the displayed license for previous package versions.

#### If your package is [open source](https://opensource.org/osd)

✔️ DO [choose an open source license](https://choosealicense.com/) to make your package open source.
> *"Open source licenses are licenses that comply with the Open Source Definition — in brief, they allow software to be freely used, modified, and shared."* - Open Source Initiative. To learn more about open source software and the Open Source Initiative, check out https://opensource.org/.

✔️ CONSIDER [including a license expression in your package](https://docs.microsoft.com/nuget/reference/msbuild-targets#packing-a-license-expression-or-a-license-file).
> License expressions are surfaced the most clearly and make it more obvious to consumers if they can use your package or if the license has changed. 
> [!Note]

> > NuGet.org only accepts license expressions that are approved by the Open Source Initiative or the Free Software Foundation.

Reviewer 8: Consider changing to:

> NuGet.org only accepts license expressions with licenses that are approved by the Open Source Initiative or the Free Software Foundation.

Because license expressions aren’t approved by OSI or FSF, only licenses.

> NuGet.org only accepts   license expressions with licenses that are approved by the Open Source   Initiative or the Free Software Foundation.
--
